src/vs/workbench/contrib/positronIPyWidgets/browser/positronIPyWidgetsService.ts

```diff
@@ -66,23 +66,39 @@ export class PositronIPyWidgetsService extends Disposable implements IPositronIP
 			this._consoleInstancesByMessageId.has(id);
 	}
 
+	/**
+	 * Map to disposeable stores for each session. Used to preventing memory leaks caused by
+	 * repeatedly attaching to the same session which can happen in the case of the application
+	 * closing before the session ends
+	 */
+	static SessionToDisposablesMap = new Map<ILanguageRuntimeSession, DisposableStore>();
+
 	private attachSession(session: ILanguageRuntimeSession) {
+		// Check if we're already attached here
+		const existingSessionDisposables = PositronIPyWidgetsService.SessionToDisposablesMap.get(session);
+		if (existingSessionDisposables && !existingSessionDisposables.isDisposed) {
+			this._logService.warn(`Already attached to session, disposing existing listeners before reattaching: ${session.metadata.sessionId}`);
+			existingSessionDisposables.dispose();
+		}
+		const disposables = new DisposableStore();
+		PositronIPyWidgetsService.SessionToDisposablesMap.set(session, disposables);
+
 		switch (session.metadata.sessionMode) {
 			case LanguageRuntimeSessionMode.Console:
-				this.attachConsoleSession(session);
+				this.attachConsoleSession(session, disposables);
 				break;
 			case LanguageRuntimeSessionMode.Notebook:
-				this.attachNotebookSession(session);
+				this.attachNotebookSession(session, disposables);
 				break;
+			default:
+				this._logService.error(`Unexpected session mode: ${session.metadata.sessionMode}`);
+				disposables.dispose();
+				PositronIPyWidgetsService.SessionToDisposablesMap.delete(session);
```

Comment:
Something I've used in cases like this, though I'm not sure if it has any downsides, is to add a disposable to `disposables` that removes the entry from the map.

Reply:
By adding the check of `.isDisposed` to the existing disposables above I was attempting to get at this, but I like the elegance (and more general usefulness) of the cleanup disposable.